src/braket/pulse/ast/free_parameters.py

```diff
@@ -11,60 +11,85 @@
-        self._expression = expression
-
-    @property
-    def expression(self) -> FreeParameterExpression:
-        return self._expression
+from oqpy.program import Program
+from oqpy.timing import OQDurationLiteral
 
 
 class _FreeParameterTransformer(QASMTransformer):
     """Walk the AST and evaluate FreeParameterExpressions."""
 
-    def __init__(self, param_values: dict[str, float]):
+    def __init__(self, param_values: dict[str, float], program: Program):
         self.param_values = param_values
+        self.program = program
         super().__init__()
 
-    def visit__FreeParameterExpressionIdentifier(
+    def visit_Identifier(
         self, identifier: ast.Identifier
-    ) -> Union[_FreeParameterExpressionIdentifier, ast.FloatLiteral]:
-        """Visit a FreeParameterExpressionIdentifier.
+    ) -> Union[ast.Identifier, ast.FloatLiteral]:
+        """Visit an Identifier.
+
+        If the Identifier is used to hold a `FreeParameterExpression`, it will be simplified
+        using the given parameter values.
+
         Args:
             identifier (Identifier): The identifier.
 
         Returns:
-            Union[_FreeParameterExpressionIdentifier, FloatLiteral]: The transformed expression.
+            Union[Identifier, FloatLiteral]: The transformed identifier.
+        """
+        if identifier.name in self.param_values:
+            return ast.FloatLiteral(float(self.param_values[identifier.name]))
+        return identifier
+
+    def visit_BinaryExpression(
+        self, node: ast.BinaryExpression
+    ) -> Union[ast.BinaryExpression, ast.FloatLiteral]:
+        """Visit a BinaryExpression.
+
+        Visit the operands and simplify if they are literals.
+
+        Args:
+            node (BinaryExpression): The node.
+
+        Returns:
+            Union[BinaryExpression, FloatLiteral]: The transformed identifier.
         """
-        new_value = identifier.expression.subs(self.param_values)
-        if isinstance(new_value, FreeParameterExpression):
-            return _FreeParameterExpressionIdentifier(new_value)
-        else:
-            return ast.FloatLiteral(new_value)
-
-    def visit_DurationLiteral(self, duration_literal: DurationLiteral) -> DurationLiteral:
-        """Visit Duration Literal.
-            node.value, node.unit (node.unit.name, node.unit.value)
-            1
+        lhs = self.visit(node.lhs)
+        rhs = self.visit(node.rhs)
+        ops = {
+            ast.BinaryOperator["+"]: operator.add,
+            ast.BinaryOperator["*"]: operator.mul,
+            ast.BinaryOperator["**"]: operator.pow,
+        }
+        if isinstance(lhs, ast.FloatLiteral):
+            if isinstance(rhs, ast.FloatLiteral):
+                return ast.FloatLiteral(ops[node.op](lhs.value, rhs.value))
+            elif isinstance(rhs, ast.DurationLiteral) and node.op == ast.BinaryOperator["*"]:
+                return OQDurationLiteral(lhs.value * rhs.value).to_ast(self.program)
```

Comment:
* Is it possible to have Duration in lhs and float in rhs?
* Are any other ops possible?